lib/pelion/Kconfig

```diff
@@ -0,0 +1,113 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+# The purpose of this file is to override and extend settings related to
+# building Pelion cloud client library.
+
+if PELION_SOURCES
+
+menu "Pelion Device Management Client Library"
+
+config MBEDTLS_CFG_FILE
+	string
+	default "config-tls-generic.h" if PELION_MBEDTLS
+
+config MBEDTLS_USER_CONFIG_FILE
+	string
+	default "config-tls-pelion.h" if PELION_MBEDTLS
+
+config MBEDTLS_SSL_MAX_CONTENT_LEN
+	int
+	default 4096
+
+config PELION_NRF_SECURITY
+	bool "Enable nrf_security module for OpenThread"
```

Comment:
This has the flaw that if user just run:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild .
$ ninja -Cbuild menuconfig
# Inside menuconfig, select `PELION_MBEDTLS=y`
# Save and exit
$ ninja -Cbuild
ninja: Entering directory `build'
[0/1] Re-running CMake...
Including boilerplate (Zephyr base (cached)): /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake
-- Application: /projects/github/ncs/nrf/applications/pelion_client
-- Zephyr version: 2.5.99 (/projects/github/ncs/zephyr)
-- Found west (found suitable version "0.10.99", minimum required is "0.7.1")
-- Board: nrf52840dk_nrf52840
-- Cache files will be written to: /home/tora/.cache/zephyr
-- Found toolchain: gnuarmemb (/opt/gcc-arm-none-eabi-7-2018-q2-update)
-- Found BOARD.dts: /projects/github/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts
-- Found devicetree overlay: /projects/github/ncs/nrf/applications/pelion_client/configuration/nrf52840dk_nrf52840/dts.overlay
-- Generated zephyr.dts: /projects/github/ncs/nrf/applications/pelion_client/build/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /projects/github/ncs/nrf/applications/pelion_client/build/zephyr/include/generated/devicetree_unfixed.h
-- Generated device_extern.h: /projects/github/ncs/nrf/applications/pelion_client/build/zephyr/include/generated/device_extern.h

warning: MBEDTLS_USER_CONFIG_ENABLE (defined at modules/Kconfig.tls-generic:334) has direct dependencies MBEDTLS_BUILTIN && MBEDTLS_CFG_FILE = "config-tls-generic.h" && MBEDTLS with value n, but is currently being y-selected by the following symbols:
 - PELION_MBEDTLS (defined at /projects/github/ncs/nrf/lib/pelion/Kconfig:68), with value y, direct dependencies PELION_SOURCES (value: y), and select condition PELION_SOURCES (value: y)

warning: MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED (defined at modules/Kconfig.tls-generic:86) has direct dependencies MBEDTLS_BUILTIN && MBEDTLS_CFG_FILE = "config-tls-generic.h" && MBEDTLS with value n, but is currently being y-selected by the following symbols:
 - PELION_MBEDTLS (defined at /projects/github/ncs/nrf/lib/pelion/Kconfig:68), with value y, direct dependencies PELION_SOURCES (value: y), and select condition PELION_SOURCES (value: y)

warning: MBEDTLS_CIPHER_AES_ENABLED (defined at modules/Kconfig.tls-generic:189) has direct dependencies MBEDTLS_BUILTIN && MBEDTLS_CFG_FILE = "config-tls-generic.h" && MBEDTLS with value n, but is currently being y-selected by the following symbols:
 - PELION_MBEDTLS (defined at /projects/github/ncs/nrf/lib/pelion/Kconfig:68), with value y, direct dependencies PELION_SOURCES (value: y), and select condition PELION_SOURCES (value: y)
```

Which happens because the nrf_security symbols that has been selected are sticky, due to them having a prompt.

This error furthermore results in the inability to re-run `menuconfig`.

Reply:
I see but this is a generic problem of kconfig. If something is automagically selected it will stay selected even if you disable the top level thing. There is nothing you could do about it.
The `PELION_NRF_SECURITY` is a convenience symbol to select symbols that Pelion would like to use.

btw - it still has OpenThread text that went unnoticed. Will remove it.

Reply:
> There is nothing you could do about it.

yes, it all depends on how you structure your own Kconfig.
Users are expecting to use menuconfig in NCS, so we should not break such a basic feature.

Reply:
next fixed. I don't think this is a bug but a limitation of a config system. there is similar case in openthread as this file was based on it

Reply:
> Users are expecting to use menuconfig in NCS, so we should not break such a basic feature.

But there is nothing that can be done here. I cannot make both options `depend on` against each other as kconfig find this as a cyclic dependency error.
The same problem is with openthread convenience symbols - if you do the same trick menuconfig fails.

Note that using menuconfig in such cases is very limited as such "convenience symbols" often enable multiple others that will not be deselected. This is just the kconfig limitation.

We could remove this but in such case there will be no possibility to set this symbols on common (lib) level as you can only do that in kconfig.